lib/devices/android/bootstrap/src/io/appium/uiautomator/core/AccessibilityNodeInfoDumper.java

```diff
@@ -0,0 +1,233 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.appium.uiautomator.core;
+import android.graphics.Point;
+import android.os.Environment;
+import android.os.SystemClock;
+import android.util.Log;
+import android.util.Xml;
+import android.view.Display;
+import android.view.accessibility.AccessibilityNodeInfo;
+import io.appium.android.bootstrap.utils.API;
+import org.xmlpull.v1.XmlSerializer;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.StringWriter;
+
+/**
+ *
+ * The AccessibilityNodeInfoDumper in Android Open Source Project contains a
+ * lot of bugs which will stay in old android versions forever. By coping the
+ * code of the latest version it is ensured that all patches become available on
+ * old android versions.
+ * 
+ * down ported bugs are e.g.
+ * { @link https://code.google.com/p/android/issues/detail?id=62906 }
+ * { @link https://code.google.com/p/android/issues/detail?id=58733 }
+ * 
+ */
+public class AccessibilityNodeInfoDumper {
+    private static final String LOGTAG = AccessibilityNodeInfoDumper.class.getSimpleName();
+    private static final String[] NAF_EXCLUDED_CLASSES = new String[] {
+            android.widget.GridView.class.getName(), android.widget.GridLayout.class.getName(),
+            android.widget.ListView.class.getName(), android.widget.TableLayout.class.getName()
+    };
+
+    /**
+     * Using {@link AccessibilityNodeInfo} this method will walk the layout hierarchy
+     * and generates an xml dump to the location specified by <code>dumpFile</code>
+     * @param root The root accessibility node.
+     * @param dumpFile The file to dump to.
+     */
+    public static void dumpWindowToFile(AccessibilityNodeInfo root, File dumpFile) {
+        final long startTime = SystemClock.uptimeMillis();
+        try {
+            FileWriter writer = new FileWriter(dumpFile);
+            XmlSerializer serializer = Xml.newSerializer();
+            StringWriter stringWriter = new StringWriter();
+            serializer.setOutput(stringWriter);
+            serializer.startDocument("UTF-8", true);
+            serializer.startTag("", "hierarchy");
+            int width = -1;
+            int height = -1;
+            if(API.API_18){
+              // getDefaultDisplay method available since API level 18
+              Display display = UiAutomatorBridge.getInstance().getDefaultDisplay();
+              Point size = new android.graphics.Point();
+              display.getSize(size);
+              width = size.x;
+              height = size.y;
+
+              serializer.attribute("", "rotation", Integer.toString(display.getRotation()));
+            }
+
+          if (root != null) {
```

Comment:
I see what's happening now.  I got thrown off by the fact that the if_API_18 logic is executed regardless of whether or not root is null.  Since you're making changes anyway, can you move the display dimension logic within the != null block?  If not possible, then no worries.